llms/mlx_lm/tuner/grpo_trainer.py

```diff
@@ -0,0 +1,680 @@
+        for i in range(0, len(idx) - batch_size + 1, batch_size):
+            yield idx[i : i + batch_size : step]
+
+    while True:
+        indices = (
+            np.random.permutation(list(batch_index_generator())) if train 
+            else batch_index_generator()
+        )
+        
+        for batch_idx in indices:
+            current_batch = [dataset[j] for j in batch_idx]
+            
+            prompts_tokens = [item[0] for item in current_batch]
+            answers_tokens = [item[1] for item in current_batch]
+            prompts_text = [item[2] for item in current_batch]
+            answers_text = [item[3] for item in current_batch]
+
+            if any(len(p) > max_seq_length for p in prompts_tokens):
+                print(
+                    f"[WARNING] Some prompts are longer than {max_seq_length} tokens. "
+                    "Long prompts will be truncated."
+                )
+
+            yield prompts_tokens, answers_tokens, prompts_text, answers_text
+
+        if not train:
+            break
+
+
+def evaluate_grpo(
+    model,
+    ref_model,
+    dataset,
+    tokenizer,
+    batch_size,
+    num_batches,
+    beta: float,
+    epsilon: float,
+    group_size: int,
+    max_seq_length,
+    reward_funcs = None,
+    loss_fn: callable = grpo_loss,
+    iterate_batches: callable = iterate_grpo_batches
+):
+    """
+    Evaluate model using GRPO loss.
+    Returns:
+        tuple: (average loss, number of tokens, average metrics)
+    """
+    all_losses = 0
+    ntokens = 0
+    all_metrics = None  # Initialize metrics dictionary
+    
+    # Create iterator for batches
+    index_iterator = iter(range(num_batches)) if num_batches != -1 else iter(int, 1)
+    
+    # Iterate through batches
+    for _, batch in zip(
+        index_iterator,
+        iterate_batches(
+            dataset=dataset,
+            tokenizer=tokenizer,
+            batch_size=batch_size,
+            max_seq_length=max_seq_length,
+        ),
+    ):
+        # Calculate loss for current batch
+        losses, toks, metrics = loss_fn(
+            model=model,
+            tokenizer=tokenizer,
+            batch=batch,
+            reward_funcs=reward_funcs,
+            beta=beta,
+            group_size=group_size,
+            epsilon=epsilon,
+            ref_model=ref_model
+        )
+        
+        # Accumulate losses and tokens
+        all_losses += losses * toks
+        ntokens += toks
+        
+        # Accumulate metrics
+        if all_metrics is None:
+            all_metrics = {k: v * toks for k, v in metrics.items()}
+        else:
+            for k, v in metrics.items():
+                all_metrics[k] += v * toks
+        
+        # Evaluate accumulated values
+        mx.eval(all_losses, ntokens)
+    
+    # Aggregate across distributed workers
+    all_losses = mx.distributed.all_sum(all_losses, stream=mx.cpu)
+    ntokens = mx.distributed.all_sum(ntokens, stream=mx.cpu)
```

Comment:
I have tried  use stream=mx.gpu and magically, my old m1 mac pro can run without OOM